Retry fed-svc creation on diff NodePort during e2e tests

clamoriniere opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#pull-request-process and developer guide https://github.com/kubernetes/community/blob/master/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://github.com/kubernetes/community/blob/master/contributors/devel/faster_reviews.md
3. Follow the instructions for writing a release note: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#release-notes
-->

**What this PR does / why we need it**:
Currently in federated end2end tests, the creation of services are
done with a randomize NodePort selection take is causing e2e test
flakes if the creation of a federated service failed if the port is
not available.
Now the util.CreateService(...) function is retrying to create the
service on different nodePort in case of error. The method retry until
success or all possible NodePorts have been tested and also failed.

**Which issue this PR fixes** 
fixes #44018

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/45186)
<!-- Reviewable:end -->

cedriclam commented:
CLA signed

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+
+		// check if service have been created in some clusters.
+		// if it's the case, delete them.
+		if found, _ := hasService(clientset, namespace, name); found {
```
Here clientset is for federation api server. I think what you wanted was to check if service is created in all federated clusters with the given NodePort and if the service creation in any one of federated clusters has failed, then probably delete that service in federation and ensure that the service is deleted in all federated cluster and create a service with another NodePort. And probably keep doing this until the service is created in all federated clusters with the chosen NodePort.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+		return false, err
 	}
-	framework.Logf("Deleting service %q in namespace %v", serviceName, namespace)
+	return true, err
```
we are returning (true, err) in case of any other errors than `IsNotFound`, which is incorrect. So please put one additional condition to check other errors.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,28 +139,82 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
-		}
-		return false, err
+		found, err := hasService(clientset, namespace, serviceName)
+		return !found, err
```
prefer to check err here.

shashidharatd commented:
@k8s-bot ok to test

cedriclam commented:
@shashidharatd thanks for your review. I updated the PR with your remarks.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,28 +139,104 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
 	err = wait.Poll(5*time.Second, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
-		_, err := clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})
-		if err != nil && errors.IsNotFound(err) {
+		found, err := hasService(clientset, namespace, serviceName)
```
I don't find any significant value in hasService function and it has made the logic little complex. Can we get rid of this function and use `clientset.Core().Services(namespace).Get(serviceName, metav1.GetOptions{})` as earlier.

cedriclam replied:
right

shashidharatd commented:
/assign @madhusudancs

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,20 +139,78 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+		needsToBeDeleted := false
+		err = wait.Poll(5*time.Second, fedframework.FederatedDefaultTestTimeout, func() (bool, error) {
+			var err error
+			_, err = clientset.Core().Services(namespace).Get(name, metav1.GetOptions{})
```
Additionally we should be checking for service shard creation in federated cluster. Refer to second point in @madhusudancs 's suggestion in https://github.com/kubernetes/kubernetes/issues/44018#issuecomment-298218566

cedriclam replied:
Your are right, 
I was thinking that calling Services(...).Delete(...) with the fedclientset.Clientset, will cascade the deletion for service shard.
I will add a loop on the ClusterSlice to call the cleanupServiceShard() function

shashidharatd commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

shashidharatd commented:
@cedriclam, sorry for the delay in checking this PR. I find some issues in the logic. PTAL at comments.

---

Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[test/e2e_federation/util.go, line 172 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r3-172:-Kjvc5WuUZq1ziO-qsRA:bpualp) ([raw file](https://github.com/kubernetes/kubernetes/blob/6abb510985f6ce62df80534de779fa1c7967f218/test/e2e_federation/util.go#L172)):*
<details><summary><i>Previously, cedriclam (Cedric Lamoriniere) wrote…</i></summary><blockquote>

Your are right, 
I was thinking that calling Services(...).Delete(...) with the fedclientset.Clientset, will cascade the deletion for service shard.
I will add a loop on the ClusterSlice to call the cleanupServiceShard() function
</blockquote></details>

Here in wait.Poll, you are checking if service is created in federation only. We should instead be checking for service shard creation in all federated clusters (in a loop for federated clusters. even if one federated cluster failed to create service, we should cleanup and try with anoter NodePort).

---

*[test/e2e_federation/util.go, line 164 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-KjvbRQVnAM9fIzA7DJ3:-KjvbRQVnAM9fIzA7DJ4:b-qpyhae) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L164)):*
> ```Go
> 		service, err := createServiceWithNodePort(clientset, namespace, name, availablePorts[i])
> 		if err == nil {
> 			return service, nil
> ```

We should not return from here. We are just successful in creating a NodePort service in federation. we still have to ensure all federated clusters are able to create the service with the same NodePort.

---

*[test/e2e_federation/util.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kjv_rKBcGw9vMwEEX1C:-Kjv_rKCTBp4GoD4-x8G:b-dqv32h) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L233)):*
> ```Go
> 	if err != nil {
> 		framework.DescribeSvc(namespace)
> 		framework.Failf("Error in deleting service %s: %v", serviceName, err)
> ```

Log the serviceName along with namespace, so that its easier to debug in case of failures.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-:-KjvcqlKrTS8TPZi4Xq1:beun67y)*
<!-- Sent from Reviewable.io -->

cedriclam commented:
@shashidharatd thanks for your review, I will update the PR with your comments this evenings.

cedriclam commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[test/e2e_federation/util.go, line 169 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r1-169:-KkSMMChylYRN_G2QhHD:bb7f2mf) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ea4860ec159959bba5ca606ca2d700e416a7fdd/test/e2e_federation/util.go#L169)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Here clientset is for federation api server. I think what you wanted was to check if service is created in all federated clusters with the given NodePort and if the service creation in any one of federated clusters has failed, then probably delete that service in federation and ensure that the service is deleted in all federated cluster and create a service with another NodePort. And probably keep doing this until the service is created in all federated clusters with the chosen NodePort.
</blockquote></details>

done

---

*[test/e2e_federation/util.go, line 172 at r3](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r3-172:-KkSMUQRW9cDo1OWkZdb:bda1wk) ([raw file](https://github.com/kubernetes/kubernetes/blob/6abb510985f6ce62df80534de779fa1c7967f218/test/e2e_federation/util.go#L172)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Here in wait.Poll, you are checking if service is created in federation only. We should instead be checking for service shard creation in all federated clusters (in a loop for federated clusters. even if one federated cluster failed to create service, we should cleanup and try with anoter NodePort).
</blockquote></details>

now the checkServicesCreation(...) is checking in all federated clusters the creation of the service.

---

*[test/e2e_federation/util.go, line 164 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-KjvbRQVnAM9fIzA7DJ3:-KkSLGFSjelYPWAa6tj9:b-c4bmsp) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L164)):*
> @shashidharatd right, code updated

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186)*
<!-- Sent from Reviewable.io -->

cedriclam commented:
Review status: 0 of 4 files reviewed at latest revision, 7 unresolved discussions.

---

*[test/e2e_federation/util.go, line 196 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r1-196:-KkSYBGlI3pIrh2s0iO-:b-qssecl) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ea4860ec159959bba5ca606ca2d700e416a7fdd/test/e2e_federation/util.go#L196)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

we are returning (true, err) in case of any other errors than `IsNotFound`, which is incorrect. So please put one additional condition to check other errors.
</blockquote></details>

fixed

---

*[test/e2e_federation/util.go, line 207 at r1](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kj4svaqgMvKvZ3-Ez7P-r1-207:-KkSOiZg10R5OJZD1vhx:b-7ql1td) ([raw file](https://github.com/kubernetes/kubernetes/blob/1ea4860ec159959bba5ca606ca2d700e416a7fdd/test/e2e_federation/util.go#L207)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

prefer to check err here.
</blockquote></details>

fixed

---

*[test/e2e_federation/util.go, line 233 at r4](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186#-Kjv_rKBcGw9vMwEEX1C:-KkSP57us_97T2Pe_f88:b-7ql1td) ([raw file](https://github.com/kubernetes/kubernetes/blob/264017c14475c06d21a5b10fc24fc38fc225fec2/test/e2e_federation/util.go#L233)):*
<details><summary><i>Previously, shashidharatd (Shashidhara T D) wrote…</i></summary><blockquote>

Log the serviceName along with namespace, so that its easier to debug in case of failures.
</blockquote></details>

fixed

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/45186)*
<!-- Sent from Reviewable.io -->

shashidharatd commented:
@madhusudancs, very strange thing happened in federation presubmit tests. it says, it still running after 6 hours. https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/45186/pull-kubernetes-federation-e2e-gce/5493/

shashidharatd commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

shashidharatd commented:
@cedriclam, there seems to be some problem with the logic. the tests are not completing. Can you please analyze logs and find the cause?

cedriclam commented:
@shashidharatd yes sure

shashidharatd commented on test/e2e_federation/authn.go:
```diff
@@ -35,16 +35,20 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
+		)
 		BeforeEach(func() {
 			fedframework.SkipUnlessFederated(f.ClientSet)
+			clusters = fedframework.ClusterSlice{}
```
here you are initializing the clusters with 0 clusters which is not the correct thing to do. In this case, the validation to see if the service was created in federated cluster would not work as it won't enter the for loop of clusters to check at all.
Instead of passing clusters from every caller. please call
```
clusters = f.GetRegisteredClusters()
```
in createService itself.

cedriclam replied:
understood

shashidharatd commented on test/e2e_federation/service.go:
```diff
@@ -102,15 +103,15 @@ var _ = framework.KubeDescribe("Federated Services [Feature:Federation]", func()
 				if service != nil {
 					By(fmt.Sprintf("Deleting service shards and their provider resources in underlying clusters for service %q in namespace %q", service.Name, nsName))
-					cleanupServiceShardsAndProviderResources(nsName, service, clusters)
+					cleanupServiceShardsAndProviderResources(nsName, service.Name, clusters)
```
we could always get `service.Name` for `service` right? why this change is required. Please try to keep your changes to minimal unless there is a strong case in doing so.

cedriclam replied:
Ok, I done this modification since the only "service *api.Service" information retrieve in the "cleanupServiceShardsAndProviderResources()" method was its name.

But sure, I will rollback those changes in order to minimalise the PR's footprint .

cedriclam commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

cedriclam commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

cedriclam commented:
Hi @shashidharatd 
I updated the PR with your last comments, but the e2e_federation test seams to have some initialisation issue... Could you help me on it?

shashidharatd commented on test/e2e_federation/authn.go:
```diff
@@ -35,16 +35,20 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
 
 			nsName := f.FederationNamespace.Name
-			svc := createServiceOrFail(fcs, nsName, FederatedServiceName)
+			svc := createServiceOrFail(fcs, nsName, FederatedServiceName, clusters)
```
@cedriclam, can you please revert passing the clusters from caller of `createServiceOrFail` and call `clusters = f.GetRegisteredClusters()` inside `createServiceOrFail` function.

cedriclam replied:
in order to call 'clusters = f.GetRegisteredClusters()' inside `createServiceOrFail`, I need to access the instance of `f *framework.Framework`. 
Also `createServiceOrFail` is a generic function used in deferent test cases, that initialise differently the  `f *framework.Framework`. 

@shashidharatd, Do you prefer that it stay like it is, or change the function signature by replacing `clusters` by `f` ? or do you see another way to revert passing `clusters` from caller of `createServiceOrFail`.

shashidharatd replied:
OK, i understand your predicament. sorry for that. its fine to pass clusters. better than the alternative

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -172,33 +226,57 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
+			cSvc = nil
+			if retry {
+				// don't return the
+				framework.Logf("Service %q in namespace %q in cluster %q not found. Trying again in %s", serviceName, namespace, name, framework.Poll)
```
broken comment

cedriclam replied:
comment update

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -172,33 +226,57 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
+			return true, err
+		}
+		By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", serviceName, namespace, name))
+		return true, err
```
shouldn't this be return true, nil?

cedriclam replied:
right.

shashidharatd commented:
@cedriclam, I am very sorry for the delay got tied up in some other priority work. will have a final look today.
Also request to squash all the commits.

cedriclam commented:
@shashidharatd, no problem for the delay, I completely understand that you can have other priority work :).

I squashed all the commits.

cedriclam commented:
@k8s-bot pull-kubernetes-unit test this

cedriclam commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

shashidharatd commented on test/e2e_federation/service.go:
```diff
@@ -53,6 +53,7 @@ var _ = framework.KubeDescribe("Federated Services [Feature:Federation]", func()
 	var _ = Describe("Without Clusters [NoCluster]", func() {
 		BeforeEach(func() {
 			fedframework.SkipUnlessFederated(f.ClientSet)
+			clusters = fedframework.ClusterSlice{}
```
we should not do this. as my earlier comment. we should get actual clusters. like `clusters = f.GetRegisteredClusters()`

cedriclam replied:
right, I forgot to make this modification that you already spot in a previous comment.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,20 +139,85 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+		if err == nil {
+			// everything was created properly so returns the federated service.
+			return service, nil
+		}
```
This block should be merged with previous block. why two `if err == nil `?

cedriclam replied:
first`ìf err == nil` is here for validation the the Service was created properly in the Federation-API-Server
the second `ìf err == nil` is here to check that the service have been creation on each federated clusters.

I didn't initially merge both, because I was considering that if `createServiceWithNodePort()` already return an error, we don't need to check on each cluster if the service was created properly.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,20 +139,85 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+
+		// in case of error, cleanup everything
+		if service != nil {
+			if err = deleteService(clientset, namespace, name, nil); err != nil {
```
why deleteService in just one cluster, when you are calling `cleanupServiceShardsAndProviderResources` 4 lines below which does cleanup all cluster services.

cedriclam replied:
I was thinking that the `clientset` is the client to target the Federation-API-server, and so not one of the federated cluster. 
That why I wanted first to delete the Service at federation level. then iterate on all the federated clusters.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,20 +139,85 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+	// we add in a array all the "available" ports
+	availablePorts := make([]int32, FederatedSvcNodePortLast-FederatedSvcNodePortFirst)
+	for i := range availablePorts {
+		availablePorts[i] = int32(FederatedSvcNodePortFirst + i)
```
This list of available ports is kind of very big. in case of some network error to some cluster. we would keep on retrying forever in the loop. So i would like it to limit to 5-10 iterations at max.

cedriclam replied:
understood, I will add a `retry`counter in order to limit the number of iterations.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -172,33 +226,57 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
+				// the function returns false in order to test again until timeout.
+				framework.Logf("Service %q in namespace %q in cluster %q not found. Trying again in %s", serviceName, namespace, name, framework.Poll)
+				return false, nil
+			}
```
introducing another parameter retry and adding a if condition is increasing the complexity of the function anyway. So i would suggest to not merge multiple scenarios into the same wait.PollImmediate function.

better to leave the original `cleanupServiceShardsAndProviderResources` intact. and use another `wait.PollImmediate` function for the other scenario.

cedriclam replied:
understood

cedriclam commented:
@k8s-bot pull-kubernetes-unit test this

cedriclam commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
-			By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", service.Name, namespace, name))
-			return true, err
-		})
+		cSvc, err := getServiceShard(namespace, service.Name, c)
```
@cedriclam  There was no need of pulling out code and creating `getServiceShard` function. Can you please revert it. I would apply the LGTM after this change. Thanks !

cedriclam replied:
ok cool!

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
+	return cSvc, err
+}
+
+func getServiceShardWithRetry(namespace string, serviceName string, cluster *fedframework.Cluster) (*v1.Service, error) {
```
Also please integrate this function `getServiceShardWithRetry` in `checkServicesCreation` itself.

cedriclam replied:
understood

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -172,33 +226,70 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
+		if err != nil && !errors.IsNotFound(err) {
+			// Get failed with an error, try again.
+			framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+			return false, nil
```
it would be better to `return false, err` here, if any err other than notfound err. Since it may be non-recoverable error and there is no point in keep on retrying.

cedriclam replied:
right, thanks for pointing it out.

cedriclam commented:
@k8s-bot pull-kubernetes-e2e-kops-aws test this

nikhiljindal commented:
/release-note-none

/approve.

Will let @shashidharatd add LGTM

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -150,20 +139,101 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
+			if err != nil && !errors.IsNotFound(err) {
+				// Get failed with an error, try again.
+				framework.Logf("Failed to find service %q in namespace %q, in cluster %q: %v. Trying again in %s", serviceName, namespace, name, err, framework.Poll)
+				return false, nil
```
@cedriclam, just the last nit. Better to return false,   err in this sentence, as any other error could potentially be non-recoverable and there is no point in waiting and retrying.

shashidharatd commented:
@cedriclam, Sorry that you had to rebase. We had to do some optimizations in the tests.
Can you please rebase this PR and address the conflicts. We could still be potentially include this PR in v1.7 as this is a test-only change.

cedriclam commented:
@shashidharatd, PR rebase done
let me know if I need to change other stuff.

cedriclam commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

cedriclam commented:
@k8s-bot pull-kubernetes-federation-e2e-gce test this

cedriclam commented:
/test pull-kubernetes-federation-e2e-gce

cedriclam commented:
@shashidharatd mind have a look to this PR? sorry for the late PR fix after rebase.

cedriclam commented:
@madhusudancs mind have a look at this PR?

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -210,18 +310,14 @@ func deleteServiceOrFail(clientset *fedclientset.Clientset, namespace string, se
 	for _, c := range clusters {
 		name := c.Name
 		var cSvc *v1.Service
-
```
unnecessary change

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -141,21 +141,84 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
 
 	By(fmt.Sprintf("Trying to create service %q in namespace %q", service.Name, namespace))
-	return clientset.Services(namespace).Create(service)
+	var err error
```
Given that we have 2 functions now `createService` and `createLBService`, checking and deleting if service shard is created in all member cluster in case of simple service is a overkill. I think the logic to check successful creation of service shard and retrying should be good enough in the scenario of `createLBService`.

cedriclam replied:
I updated the createService function as you proposed: create federated service, then check creation.

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -237,7 +333,6 @@ func cleanupServiceShardsAndProviderResources(namespace string, service *v1.Serv
 			By(fmt.Sprintf("Service %q in namespace %q in cluster %q found", service.Name, namespace, name))
 			return true, err
 		})
-
```
unnecessary change

shashidharatd reviewed (changes requested):
@cedriclam, I am extremely sorry for such a long delay, as i was not well and had taken a break. I wish we could address the comments in this PR and close it within couple of days.
The concern i have is to minimize the test execution time in case of failures and would like to touch minimal code to make these changes.
Thanks a lot for holding up.

cedriclam commented:
/test pull-kubernetes-e2e-kops-aws

cedriclam commented:
/test pull-kubernetes-e2e-kops-aws

cedriclam commented:
@shashidharatd now worries for the delay. I updated the PR with your comments.
Let me know if it is looking good for you, and if you see other changes needed.
Thanks for your time.

cedriclam commented:
/test all

cedriclam commented:
/retest

shashidharatd commented on test/e2e_federation/util.go:
```diff
@@ -141,21 +141,71 @@ func createService(clientset *fedclientset.Clientset, namespace, name string) (*
 
 	By(fmt.Sprintf("Trying to create service %q in namespace %q", service.Name, namespace))
-	return clientset.CoreV1().Services(namespace).Create(service)
+	var err error
```
This block may not be required and the original one should suffice. Can you please revert this one and also preserve the signature of createService function

cedriclam replied:
sure

clamoriniere committed:
Retry fed-lb-svc creation on diff NodePort during e2e tests

Currently, in federated end2end tests, the creation of services are
done with a randomize NodePort selection. It causing e2e test
flakes if the creation of a federated service failed if the port is
not available.
Now the util.CreateService(...) function is re trying to create the
service on different nodePort in an error case. The method retries until
success or 10 creation retry with other random NodePorts.
If never the service has not been created properly on one of the
federated cluster, a Service shards cleanup is executed before retrying
again the federated service creation.

fixes #44018

shashidharatd commented:
/retest

shashidharatd commented:
/lgtm
